src/compute/src/sink/copy_to_s3_oneshot.rs

```diff
@@ -77,18 +80,39 @@ where
             "Consolidated COPY TO S3 input",
         );
 
+        // We need to consolidate the error collection to ensure we don't act on retracted errors.
         let error = consolidate_pact::<KeyBatcher<_, _, _>, _, _, _, _, _>(
             &err_collection.map(move |row| {
                 let batch = row.hashed() % batch_count;
                 ((row, batch), ())
             }),
             Exchange::new(move |(((_, batch), _), _, _)| *batch),
-            "Consolidated COPY TO S3 error",
+            "Consolidated COPY TO S3 errors",
         );
+        // We can only propagate the one error back to the client, so filter the error
+        // collection to the first error that is before the sink 'up_to' to avoid
+        // sending the full error collection to the next operator.
+        let error_stream =
+            error
+                .inner
+                .unary_frontier(Pipeline, "COPY TO S3 error filtering", |_cap, _info| {
+                    let up_to = sink.up_to.clone();
+                    let mut vector = Vec::new();
+                    let mut received_one = false;
+                    move |input, output| {
+                        while let Some((time, data)) = input.next() {
+                            if !up_to.less_equal(time.time()) && !received_one {
+                                received_one = true;
+                                data.swap(&mut vector);
+                                output.session(&time).give_iterator(vector.drain(..1));
+                            }
+                        }
+                    }
```

Comment:
we do this `up_to` check on the input side right? Considering that, it seems like this could be:
```
let single_error = false;
let error_stream = error.inner.filter(|_| {
	if !single_error {
		single_error = true; 
		true
	} else {
		false
	}
});
```

Reply:
We do the `up_to` check downstream of this in the sink initialization operator - so I was concerned that if we naively filter to the first of any error here, and that error is beyond `up_to` we might miss a potential first error from before `up_to`. does that make sense?

Reply:
ah because of the ordering...makes sense! 

probably worth a comment!